python/dgl/sparse/sparse_matrix.py

```diff
@@ -194,6 +171,190 @@ def transpose(self):
         """
         return SparseMatrix(self.c_sparse_matrix.transpose())
 
+    def to(self, device=None, dtype=None):
+        """Perform matrix dtype and/or device conversion. If the target device
+        and/or dtype are already in use, the original matrix will be returned.
+
+        Parameters
+        ----------
+        device : torch.device, optional
+            This is the target device of the matrix. If not given, the current
+            device will be used.
+        dtype : torch.dtype, optional
+            This is the target data type of the matrix values. If not given,
+            the current data type will be used.
+
+        Returns
+        -------
+        SparseMatrix
+            The result matrix
+
+        Example
+        --------
+
+        >>> row = torch.tensor([1, 1, 2])
+        >>> col = torch.tensor([1, 2, 0])
+        >>> A = from_coo(row, col, shape=(3, 4))
+        >>> A.to(device='cuda:0', dtype=torch.int32)
+        SparseMatrix(indices=tensor([[1, 1, 2],
+                                     [1, 2, 0]], device='cuda:0'),
+                     values=tensor([1, 1, 1], device='cuda:0',
+                                   dtype=torch.int32),
+                     size=(3, 4), nnz=3)
+        """
+        if device is None:
+            device = self.device
+        if dtype is None:
+            dtype = self.dtype
+
+        if device == self.device and dtype == self.dtype:
+            return self
+
+        row, col = self.coo()
```

Comment:
Use val_like instead of getting coo explicitly. COO may not be materialized.

Reply:
This only works if the device stays unchanged.